### ISSUE TITLE
OpenGL texture display in the Inputs panel is confusing

### ISSUE THREAD

**Issue by redorav** [baldurk/renderdoc#711]
I've been looking at a shader quite puzzled for a bit of time, only to realize that the OpenGL backend presents the textures in the Inputs panel alphabetically. It assigns indices based on this that say FS0, FS1, etc which crucially do not match the slot it's actually bound to. Is there some setting I've turned on by accident that'll do this or is this a bug?

Thanks!

**Comment by baldurk** [comment]
I'm not entirely sure I understand the problem.

To clarify understanding, for OpenGL you have texture bindings that have some arbitrary implementation defined locations. These values don't really mean anything they just have to be queried (if you use them). Let's be nice, and number them consecutively from 0:

```
location 0: texA
location 1: texB
location 2: myTex
```

The uniforms in these locations then contain a single integer, which indicates which texture binding is in use. These days this is usually specified in the shader with `layout()`, but even then it can be changed later:

```
location 0: texA    --> texture slot 3
location 1: texB    --> texture slot 5
location 2: myTex   --> texture slot 4
```

Then these texture slots have the actual textures bound, which on GL are also uints, but renderdoc prefers to use names:

```
location 0: texA    --> texture slot 3 --> Tex2D 583
location 1: texB    --> texture slot 5 --> Diffuse Brick Texture
location 2: myTex   --> texture slot 4 --> Tex3D 99
```

So on the GL pipeline state view, RenderDoc will show the textures in their location order ( [\*] sort of - see below), show the slot that they're pointing to, and then the texture that's bound.

```
3: texA    | Tex2D 583
5: texB    | Diffuse Brick Texture
4: myTex   | Tex3D 99
```

In the texture viewer, I drop some of that complexity for the sake of readability and usability. The thumbnails are named in location order pointing to the texture:

```
FS0: texA  = Tex2D 583
FS1: texB  = Diffuse Brick Texture
FS2: myTex = Tex3D 99
```

There are two reasons for this: one thing, it smooths out some of the differences between APIs in a case where people might not care about it and there's value in having unified behaviour. Binding models are a place where the APIs differ a lot, so while no other API has this kind of indirection scheme that GL does, every API has the concept of "The first read-only resource declared in the shader, the second read-only resource declared, etc.".

The other thing is kind of related, but those thumbnails are used for tracking which API slot you're looking at if you're moving around events. So when you select the first colour target output (which is the default), as you step around then the tab shows you whatever is bound there. For the input tabs though this becomes more complex, so tracking 'the first read-only resource' or 'the third read-only resource' is always consistent, whereas e.g. for GL there is no other consistent value, since the slot could change depending on the uniform value as it's mutable state.

Hopefully that make sense. Part of the problem here I think is that GL has this awful level of indirection with both the slot and the texture binding both being mutable, that is (AFAIK) in practice rarely used.

[\*] I mentioned location order above, this is one area where essentially random-ness comes in. The GL resources are reflected out of the shaders using the program introspection extension, but it's implementation dependent what order these come back from those queries. I believe some drivers return them in declaration order (which feels natural), and some of them return them in alphabetical order.

There's no alphabetical sort applied by RenderDoc, so perhaps this is the real source of your problem, that the driver ends up sorting the resources it reports and RenderDoc displays in that order?

**Comment by redorav** [comment]
Of course, that makes perfect sense. Is there a way then to pair them together for display, so say (location = 0, slot = 3) or something along those lines that'll be less confusing?

I was comparing two captures, one with OpenGL and another with DX11, which is where this issue presented itself. In DX11 the ordering of textures in the pipeline view is the same as the ordering in the Inputs view, where I can quickly visually inspect if all my textures are bound correctly, etc.

In OpenGL I can't do that as easily because like you say all the textures are sorted in whatever order the driver happened to assign the slots, and it's easy to get confused about it. There isn't also an easy way to see (from the Inputs tab) the slot it's assigned to, or the location<->slot mapping that's being performed. In our case, we let the driver to the location assignment, query it and tell it which slot we want instead of using the location qualifier to tell it where to go so we don't much care about it, but we care about the slot.

**Comment by baldurk** [comment]
I don't _think_ I agree with what you're suggesting, though I might be misunderstanding where you're coming from.

One thing you say:

> In DX11 the ordering of textures in the pipeline view is the same as the ordering in the Inputs view,

sounds like you might be describing a bug. This should be true for DX11, but it should be true for OpenGL as well. The inputs in both cases should always be listed in essentially "the order of resources that are reflected from the shader", so if that's not the case for GL then it's a bug I'd like to fix.

The only difference is that the texture viewer's inputs tab labels the items with essentially the row index from the pipeline state view, and the state viewer shows the actual texture slot and texture used - which would be the `3: texA, 5: texB, 4: myTex` in my example above.

DX11 does act a bit strange due to legacy reasons that I plan to clean up some time, since it predates renderdoc knowing how to handle multiple APIs. Every DX11 shader acts like it has 128 textures bound, but some (most!) of them are marked 'unused'. This leads to the mapping being an identity and so the 7th texture is also kind of coincidentally always at slot 6, and has the side effect of showing the binding slot in the thumbnail because DX11 is 'simple' and only has a single fixed shader-side binding with no mutable state or indirections. This will change in future to act the same way as the other APIs, and it's not directly comparable to GL where there is an extra mutable indirection.

I don't want to display the slot numbers as well as the 'resource order index' in the texture viewer thumbnails, because I feel that's kind of the wrong place for that information and will just clutter without adding much value. In the case of D3D12 for example (which has an awful binding model) to be equivalent you'd need to display the index in the root signature, the register space, and the register, as well as the resource index. This is a lot of information to display on the thumbnail which in the majority of cases is not useful. Since it is a thumbnail, space is extremely precious and I need to weigh carefully what's displayed there.

To me it's about a division of labour. The texture viewer is about displaying which textures are bound and inspecting their contents visually. The thumbnails are there to browse to the different resources that are bound, not to inspect the details of the binding information. For that you have the pipeline state view which has more space to give further detailed information in a more fitting format.

btw I don't think you're asking to display the actual GL uniform locations, but FWIW I also don't want to those since I think that again adds a lot of mess and confusion into the UI without any value.

**Comment by redorav** [comment]
I see where you're going, and I guess in my mind I'm always thinking in DX11 "slot" terms, but of course it's going to be even less simple in the future APIs. Let me simply state write an example of what I'm seeing in each API and you tell me if perhaps my expectation is wrong.

```
DX11:

Pipeline View
3 : texC, 5 : myTex, 7 : texA

Inputs
PS3 texC
PS5 myTex
PS7 texA
```

```
OpenGL:

Pipeline View
3 : texC, 5 : myTex, 7 : texA

Inputs
FS0 myTex
FS1 texA
FS2 texC
```

This is what confuses me, because Inputs in GL refers to locations in the shader whereas pipeline view refers to slot indices. They appear the same in DX11 and I can make a correlation between both, whereas in OpenGL I need to do a sort of mapping in my mind, and the textures aren't even sorted in the same order as I've bound them in the Pipeline.

If it's simpler I can even take a few screenshots. I'm not at work anymore but tomorrow I can take 4 screenshots, one for the pipeline and one for the texture viewer, for both dx11 and opengl from a real example.

**Comment by baldurk** [comment]
OK so in that case, the expected result for me today would be slightly tweaked:

```
OpenGL:

Pipeline View
3 : texC, 5 : myTex, 7 : texA

Inputs
FS0 texC
FS1 myTex
FS2 texA
```

Or possibly

```
OpenGL:

Pipeline View
5 : myTex, 7 : texA, 3 : texC

Inputs
FS0 myTex
FS1 texA
FS2 texC
```

depending on which is the 'true' reflected order of the samplers.

That would be confusing having them listed in non-ascending order, but having two different orders for the bindings that can disagree means that in the event of a discrepancy something's going to lose out. It sucks but is just kind of inevitable in GL. The reflected shader order should 'win' since that should be fixed, regardless of any current state

Later when I fix DX11 to work consistently, the DX11 one would change to:

```
DX11:

Pipeline View
3 : texC, 5 : myTex, 7 : texA

Inputs
PS0 texC
PS1 myTex
PS2 texA
```

So it sounds like the key thing that you're running into is that the inputs are not being displayed in a consistent order, and that's a bug. They should be listed in the order they come out of the shader reflection, even if that means that the actual slots aren't sorted (since you can set them to anything arbitrarily). This might be to do with the way bindings are iterated in the pipeline state view that inadvertently sorts by the slot when it shouldn't.

If you can share a capture that exhibits this that would be ideal, but otherwise I can try and see if I can reproduce it here (I'm not entirely sure what produces that behaviour and it might be dependent on driver, but I can try and poke around). Either way a fix for this is unlikely to make it into the next release though I'm afraid but at least if I find a fix then you can apply it and build locally.

**Comment by redorav** [comment]
Sure, I'm not in a hurry and I can work around it, if you have a fix for me locally I'd be happy to test it out.

From your comments it seems like my expectation was a little off but thinking about it, it makes sense that the shader determines the display order and the pipeline just reflects what you've bound where with respect to the shader. As long as it's consistent I'd be happy with the proposed change (even if slots are "swapped").

I'll try to send you captures tomorrow although it might not be possible, but I'll send you anything I can to help you out.

**Comment by redorav** [comment]
I've sent you an email with two captures that can help you out.

**Comment by baldurk** [comment]
I prototyped the way I'd outlined but it just looked way too weird to constantly have the resources be listed in (apparently) unsorted/random order. It's "pure" in concept but unhelpful in practice.

The better compromise that I went with was to acknowledge that although GL technically allows you to remap these bindings at any time, for any sensible codebase they are fixed - either with `layout()` statements in the shaders themselves, or via a fixed deterministic mapping set with reflection like you have. In those cases the binding order is useful and fixed, and it should be used in preference.

This does mean I now have to modify the shader reflection data which should really be compile-time immutable, re-ordering the bindings based on the uniform values every time. It's ugly but it's relatively contained, and it gives you the intuitive ordering you expect.

I've only committed this onto the `v1.x` branch though, as I'm not yet sure it's bug and edge-case free. If it proves to be stable then I'll cherry-pick it back to `v0.x`. You're free to apply the commit yourself onto `v0.x` if you want (that code hasn't diverged so it should apply cleanly).
